### ISSUE TITLE
[cs] List_1.ConvertAll gives false positive (warning)

### ISSUE THREAD

**Issue by kLabz** [HaxeFoundation/haxe#8589]
The following code results in a `This expression may be invalid` warning on the second line:
```haxe
var list:List_1<T> = new List_1();
list.ConvertAll((t:T) -> t.id);
```

When printing the unification error in `CastDetect.infer_params` we get:
```
cs.system.Converter_2<T, String> should be System.Converter`2<T, String>
```

To reproduce:
```haxe
// Issue8589/Main.hx
import cs.system.collections.generic.List_1;

typedef T = {id:String};

class Main {
        public static function main() {
                var list:List_1<T> = new List_1();
                list.ConvertAll((t:T) -> t.id);
        }
}
```

```hxml
# Issue8589/compile.hxml 
-cs bin
-main Main
-D no-compilation
```

**Comment by RealyUniqueName** [comment]
I debugged this for a bit.
It indeed looks like at some point after `gencommon` transformations unrelated types get unified.
The error in that sample comes from 
https://github.com/HaxeFoundation/haxe/blob/6138b295a2deedf50a1b953b8a05e5fa87e8fff0/src/codegen/gencommon/castDetect.ml#L703-L715
It didn't happen in Haxe 3.4, but is already reproducible with 4.0-rc.3. 
Didn't check versions inbetween. So, the next step is to bisect the commit, which caused that.

**Comment by kLabz** [comment]
First rough "bisect": it happened between preview 4 and preview 5 (yeah that's only 490 commits o/)

Current range: https://github.com/HaxeFoundation/haxe/compare/1e3e5e0...45d662c

**Comment by kLabz** [comment]
@RealyUniqueName the warning goes away when reverting 0c8ce47bb0ffff82af3151a969315df1cff4f9cf. Not sure how to fix #6635

**Comment by RealyUniqueName** [comment]
Thanks! I'll continue from there

**Comment by kLabz** [comment]
Any lead on what would need to be done instead of 0c8ce47 to fix #6635?
